### PR TITLE
Webwork Script: Create Markdown Files using Webwork Questions

### LINKED ISSUE #6: Content from Webwork on GitHub
I had a meeting with Agnes D'Entremont and Jennifer Kirkey from UBC-V and Douglas College and they have put together a LOT of physics problems.

Here is the email in full, along with the links:

> Okay, here’s the content dump.  Feel free to ask about any of this. 
> 
> What we’ve done: 
> Our 2nd year engineering problems (multiple topics): https://github.com/ubc-mech2/UBC-engineering-webwork
> (Note that “tested” means deployed in a class. Need to deploy before submission to OPL. All problems tested by development team).  
> 
> Our 1st and 2nd year mechanics problems (working directory): https://github.com/ubc-mech2/OER-mechanics-webwork
> 
> 2nd year biomechanics: https://github.com/ubc-mech2/OER-biomechanics-webwork
> 
> In the above repos, any images without a CC license shouldn’t be used. 
> 
> 
> What others have done:
> OpenStax Physics: https://github.com/openwebwork/webwork-open-problem-library/tree/master/Contrib/BrockPhysics
> 
> Statics: https://github.com/openwebwork/webwork-open-problem-library/tree/master/Contrib/WhatcomCC
> 
> Statics, Thermo, Solid Mech: https://github.com/openwebwork/webwork-open-problem-library/tree/master/Contrib/LaTech
> 
> Fluids: https://github.com/openwebwork/webwork-open-problem-library/tree/master/Contrib/QueensU
> 
> EE: https://github.com/openwebwork/webwork-open-problem-library/tree/master/Contrib/NCSU
> 
> Cheers,
> Agnes
>

### PR DESCRIPTION
**Pull request by Parsa-Rajabi**
# Change Log
- Created `Webwork` directory
- Created `webwork_to_md.py` script 

# Github Issue
closes #6

### CODE REVIEW

**Comment by cbm755** [comment]
Hi @firasm et al, I tried this on two semi-randomly selected WW from OPL.  It fails b/c my .pg files did not contain `BEGIN_HINT`.

```
Traceback (most recent call last):
  File "/home/cbm/work/teaching/math253/2021s1/tmp/webwork_to_md.py", line 45, in <module>
    problem_full = file_contents[file_contents.index(start_of_problem_src):file_contents.index(begin_hint_src)]
ValueError: substring not found
```

If you want to reproduce, to files I tried were:
```
FortLewis/Calc3/12-1-Two-variable-functions/HGM4-12-1-02-Functions-of-two-variables.pg
272/setStewart12_1/problem_1.pg
```

**Comment by firasm** [comment]
Thanks @cbm755 ! Looks like all of the physics ones we were looking at had hints. Still a work-in-progress so hopefully we have a cleaned up version soon! 

cc: @Parsa-Rajabi

File: webwork/webwork_to_md.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+problem_clean_up = re.sub(r"ans_rule(.+?[(])(.+?[)])", '', problem_multi_para)
+# remove ANS line from problem
+problem_no_ans = re.sub(r"ANS(\(.+?[?<!)]\));", '', problem_clean_up)
+# remove empty lines from problem
+problem_no_empty_lines = os.linesep.join([s for s in problem_no_ans.splitlines() if s])
+# remove hint lin from beginning of problem
+problem_text = problem_no_empty_lines[problem_no_empty_lines.index("hint.")+6:]
+
+# get answer from file
+start_of_answer_src = "showHint"
```
Comment (cbm755) [inline webwork/webwork_to_md.py:69]: my src does not end with this

File: webwork/webwork_to_md.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+end_of_answer_src = "BEGIN_TEXT"
+
+answer_with_hint = file_contents[file_contents.index(start_of_answer_src):file_contents.index(end_of_answer_src)]
+answer_section = re.sub(rf"{start_of_answer_src} =\d+;", '', answer_with_hint)\
+    .replace(end_of_answer_src, '')\
+    .replace(';', '')\
+    .replace('$', '').strip()
+if "random" in answer_section:
+    answer_section = "from random import randrange\n" + answer_section.replace('random', 'randrange')
```
Comment (cbm755) [inline webwork/webwork_to_md.py:78]: Here I suggest using a regex matching "variable = random(...)".  Current this is too greedy.  For example, file I'm looking at has:
```
$x0 = random(-10,10);
$y0 = non_zero_random(-10,10);
$z0 = non_zero_random(-10,10);
```

A more conservative approach is explicitly match lines.  And lines that are not explicitly and conservatively translated get comments like ` # TODO port me` appended.

**Comment by Parsa-Rajabi** [comment]
> Hi @firasm et al, I tried this on two semi-randomly selected WW from OPL. It fails b/c my .pg files did not contain `BEGIN_HINT`.
> 
> ```
> Traceback (most recent call last):
>   File "/home/cbm/work/teaching/math253/2021s1/tmp/webwork_to_md.py", line 45, in <module>
>     problem_full = file_contents[file_contents.index(start_of_problem_src):file_contents.index(begin_hint_src)]
> ValueError: substring not found
> ```
> 
> If you want to reproduce, to files I tried were:
> 
> ```
> FortLewis/Calc3/12-1-Two-variable-functions/HGM4-12-1-02-Functions-of-two-variables.pg
> 272/setStewart12_1/problem_1.pg
> ```

Thanks @cbm755 for the review. Could you please provide a link to those math pg files?

File: webwork/webwork_to_md.py (hunk 2 above)
Comment (Parsa-Rajabi) [reply webwork/webwork_to_md.py:78]: Thanks for the suggestions! The answer section is still WIP but I will definitely keep this in mind :)

**Comment by cbm755** [comment]
those are the links with OLP, do you need a direct link?

**Comment by firasm** [comment]
> those are the links with OLP, do you need a direct link?

Hey @cbm755 I tried to find that problem here: https://github.com/openwebwork/webwork-open-problem-library/tree/master/OpenProblemLibrary/FortLewis/Calc3/12-1-Two-variable-functions

But that directory doesn't seem to exist...

**Comment by cbm755** [comment]
It works for me...   Here are the two direct links:

https://github.com/openwebwork/webwork-open-problem-library/blob/master/OpenProblemLibrary/FortLewis/Calc3/12-1-Two-variable-functions/HGM4-12-1-02-Functions-of-two-variables.pg

https://github.com/openwebwork/webwork-open-problem-library/blob/master/OpenProblemLibrary/272/setStewart12_1/problem_1.pg

I think maybe you're trying to combine my two examples into one or something

**Comment by firasm** [comment]
Ha! Silly me... sorry about that! I had appended the second one to the first one.

**Comment by Parsa-Rajabi** [comment]
@cbm755 I've updated the script to handle problems with no hint (although most physics problems do have hints). Feel free to try the script on some math problems :) (answer section is still WIP).

**Comment by Parsa-Rajabi** [comment]
webwork formatting to be clarified via #21

**Comment by firasm** [comment]
Is this ready for me to review @Parsa-Rajabi ?

**Comment by Parsa-Rajabi** [comment]
closing to rebase

**Pull request closed without merging**